### PR TITLE
Blog post6 grids

### PR DESCRIPTION
**Pull request by dpaquette**
Finally finished this monster. The Updating Complex Grids post will probably be split into 2 posts for CDN Devs but can stay together for the book.

### CODE REVIEW

File: Blog 6 - Updating Simple Grid Components.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+		case when isnull(pj_auto_subscribe_default_user,0) = 1 then 'Y' else 'N' end [auto subscribe<br>default user],
+		case when isnull(pj_enable_pop3,0) = 1 then 'Y' else 'N' end [receive items<br>via pop3],
+		pj_pop3_username [pop3 username],
+		pj_pop3_email_from [from email address],
+		case when pj_default = 1 then 'Y' else 'N' end [default]
+		from projects
+		left outer join users on us_id = pj_default_user
+		order by pj_name"));
+```
+Notice how there is HTML embedded inside this query. In order to have display edit/delete links, we need the cells in the data set to contain the actual HTML that will be rendered on the client.
```
Comment (stimms) [inline Blog 6 - Updating Simple Grid Components.md]: are we also mixing data and view level concerns?

File: Blog 6 - Updating Complex Grid Components.md [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+
+Given the current implementation, we can safely assume that the BugTracker deployments do not typically contain thousands of active bugs or thousands of active users. Regardless, I would like to aim for better performance with the bugs grid. The current implementation is also difficult to understand. Let's see if we can both improve the performance and maintainability of this code.
+
+#Filtering, Sorting and Paging in SQL
+One of the biggest challenges here is the fact that this page can execute arbitrary queries that return any number of columns. This seems to be an important extensibility point in BugTracker so I would like to keep this feature.
+
+If we want to handle filtering, sorting and paging on the database server, then we will to extend the BugTracker queries feature.
+
+The base installation of BugTracker contains about 10 different queries that can be executed on the bugs.aspx page. These can be added to by administrators, but they all look a little something like this:
```
Comment (stimms) [inline Blog 6 - Updating Complex Grid Components.md]: I worry that we might be mixing messages here. I think I've come out pretty strongly against allowing these sorts of custom queries in the the past. It is dangerous to let users enter their own SQL.

File: Blog 6 - Updating Complex Grid Components.md [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+        _columnNames = query.ColumnNames;
+    }
+
+    public BugQueryResult ExecuteQuery(IIdentity identity, int start, int length, string orderBy,
+        string sortDirection, BugQueryFilter[] filters = null)
+    {
+        return ExecuteQuery(identity, start, length, orderBy, sortDirection, false, filters);
+    }
+
+    public BugQueryResult ExecuteQuery(IIdentity identity, int start, int length, string orderBy, string sortDirection, bool idOnly, BugQueryFilter[] filters = null)
```
Comment (stimms) [inline Blog 6 - Updating Complex Grid Components.md]: This feels like a long function

File: Blog 6 - Updating Complex Grid Components.md (hunk 2 above)
Comment (dpaquette) [inline Blog 6 - Updating Complex Grid Components.md]: The users aren't allowed to enter SQL anymore, we (you) took that feature out. The queries can only be customized by a developer or admin who has access to the database and hopefully knows what they are doing. I think that is consistent with what you said in your Revealing Sensitive Data post:
>Let's take away the ability to edit queries but leave intact the actual query functionality. This means that people can still create the queries on the back end should they be needed.

Comment (stimms) [inline Blog 6 - Updating Complex Grid Components.md]: Okay, perfect. I was pretty sure I took that out but couldn't remember. I'm old like @canadianjames

File: Blog 6 - Updating Complex Grid Components.md [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+
+###Setting up the Environment
+The first step is to create an environment were we can compare the performance of the old and new implementations. To do this, I switched back to the *master* branch to get a copy of the BugTracker with the original data grid implementation. I took a compiled version of this and deployed it as a web application on my local IIS. I called this application *btnetold* and assigned it to a new application pool by the same name. Next, I switched back to the *Data_Grid_Update* branch to get a copy of the new implementation. I compiled this version and deployed it to a web application called *btnetnew*, also with a new application pool by the same name. I now have the old and new applications installed into IIS. I am able to turn each version on/off by starting and stopping the individual application pools. Both instances are configured to use the same database instance which is also hosted on my local machine.
+
+![Original and new implementations installed locally](Images/LocalIISDeployments.png)
+
+###Designing the Tests
```
Comment (stimms) [inline Blog 6 - Updating Complex Grid Components.md]: This whole section is great. Learned a lot.

File: Blog 6 - Updating Complex Grid Components.md (hunk 3 above)
Comment (dpaquette) [inline Blog 6 - Updating Complex Grid Components.md]: It's 43 lines. Do you have any suggestions on how to make it shorter and more readable?

Comment (stimms) [inline Blog 6 - Updating Complex Grid Components.md]: If this was a real system I would say fix it.  However adding proper layers of abstraction would require a bunch more classes which would be difficult to show. I say leave it.

**Pull request merged**